action.yml

```diff
@@ -1,35 +1,134 @@
   color: blue
 
 runs:
   using: composite
   steps:
+    - name: "Validate input"
+      id: validate-input
+      run: |
+        def filter_version(version):
+            """return python 'major.minor'"""
+            if version.startswith("pypy-"):
+                version_ = version[5:]
+            elif version.startswith("pypy"):
+                version_ = version[4:]
+            else:
+                version_ = version
+            version_ = version_.split("-")[0]  # remove extra specifier e.g. "3.11-dev" => "3.11"
+            version_parts = version_.split(".")
+            if len(version_parts) < 2:
+                raise ValueError(f"invalid version: {version}")
+            if not version_parts[0].isdigit():
+                raise ValueError(f"invalid major python version: {version}")
+            if not version_parts[1].isdigit():
+                raise ValueError(f"invalid minor python version: {version}")
+            return ".".join(version_parts[:2])
+
+
+        input = "${{ inputs.python-versions }}"
+        versions = [version.strip() for version in input.split(",")]
+
+        pypy_versions = [version for version in versions if version.startswith("pypy")]
+        pypy_versions_filtered = [filter_version(version) for version in pypy_versions]
+        if len(pypy_versions) != len(set(pypy_versions_filtered)):
+            raise ValueError(f"multiple versions specified for the same 'major.minor' PyPy interpreter: {pypy_versions}")
+
+        cpython_versions = [version for version in versions if version not in pypy_versions]
+        cpython_versions_filtered = [filter_version(version) for version in cpython_versions]
+        if len(cpython_versions) != len(set(cpython_versions_filtered)):
+            raise ValueError(f"multiple versions specified for the same 'major.minor' CPython interpreter: {cpython_versions}")
+
+        # cpython shall be installed last because
+        # other interpreters also define pythonX.Y symlinks.
+        versions = pypy_versions + cpython_versions
+
+        if "3.10" in cpython_versions_filtered:
+            # need to place this last
+            index = cpython_versions_filtered.index("3.10")
+            index = versions.index(cpython_versions[index])
+            cpython_310 = versions.pop(index)
+            versions.append(cpython_310)
+        else:
+            # add this to install nox
+            versions.append("3.10")
+
+        if len(versions) > 15:
+            raise ValueError(f"Too many interpreters to install: {len(versions)} > 15")
+
+        print(f"::set-output name=interpreter_count::{len(versions)}")
+        for i, version in enumerate(versions):
+            print(f"::set-output name=interpreter_{i}::{version}")
+      shell: "${{ runner.os == 'Windows' && 'python' || 'python3' }} {0}"
     - uses: actions/setup-python@v3
       with:
-        python-version: "pypy-3.7"
+        python-version: "${{ steps.validate-input.outputs.interpreter_0 }}"
```

Comment:
Apologies for my flip flopping on this but if this level of additional work is what's required to avoid the `"['3.7', '3.8', '3.9'...]"` syntax then I don't think it's worth it and I'd be happier to live with the string list. At least with the former we don't have to have a custom validator?